pkg/smi/client_test.go

```diff
@@ -0,0 +1,585 @@
+	smiTrafficTargetClientSet := testTrafficTargetClient.NewSimpleClientset()
+	osmPolicyClientSet := osmPolicyClient.NewSimpleClientset()
+	namespaceController := namespace.NewNamespaceController(kubeClient, meshName, stop)
+
+	fakeClientSet := &fakeKubeClientSet{
+		kubeClient:                kubeClient,
+		smiTrafficSplitClientSet:  smiTrafficSplitClientSet,
+		smiTrafficSpecClientSet:   smiTrafficSpecClientSet,
+		smiTrafficTargetClientSet: smiTrafficTargetClientSet,
+		osmPolicyClientSet:        osmPolicyClientSet,
+	}
+
+	// Create a test namespace that is monitored
+	testNamespace := corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   testNamespaceName,
+			Labels: map[string]string{namespace.MonitorLabel: meshName}, // Label selectors don't work with fake clients, only here to signify its importance
+		},
+	}
+	if _, err := kubeClient.CoreV1().Namespaces().Create(context.TODO(), &testNamespace, metav1.CreateOptions{}); err != nil {
+		log.Fatal().Err(err).Msgf("Error creating Namespace %v", testNamespace)
+	}
+	<-namespaceController.GetAnnouncementsChannel()
+
+	meshSpec, err := newSMIClient(
+		kubeClient,
+		smiTrafficSplitClientSet,
+		smiTrafficSpecClientSet,
+		smiTrafficTargetClientSet,
+		osmPolicyClientSet,
+		osmNamespace,
+		namespaceController,
+		kubernetesClientName,
+		stop,
+	)
+
+	return meshSpec, fakeClientSet, err
+}
+
+var _ = Describe("When listing TrafficSplit", func() {
+	var (
+		meshSpec      MeshSpec
+		fakeClientSet *fakeKubeClientSet
+		err           error
+	)
+	BeforeEach(func() {
+		meshSpec, fakeClientSet, err = bootstrapClient()
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	It("should return a list of traffic split resources", func() {
+		split := &smiSplit.TrafficSplit{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test-ListTrafficSplits",
+				Namespace: testNamespaceName,
+			},
+			Spec: smiSplit.TrafficSplitSpec{
+				Service: tests.BookstoreApexServiceName,
+				Backends: []smiSplit.TrafficSplitBackend{
+					{
+						Service: tests.BookstoreServiceName,
+						Weight:  tests.Weight,
+					},
+				},
+			},
+		}
+
+		_, err := fakeClientSet.smiTrafficSplitClientSet.SplitV1alpha2().TrafficSplits(testNamespaceName).Create(context.TODO(), split, metav1.CreateOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		<-meshSpec.GetAnnouncementsChannel()
+
+		splits := meshSpec.ListTrafficSplits()
+		Expect(len(splits)).To(Equal(1))
+		Expect(split).To(Equal(splits[0]))
+
+		err = fakeClientSet.smiTrafficSplitClientSet.SplitV1alpha2().TrafficSplits(testNamespaceName).Delete(context.TODO(), split.Name, metav1.DeleteOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		<-meshSpec.GetAnnouncementsChannel()
+	})
+})
+
+var _ = Describe("When listing TrafficSplit services", func() {
+	var (
+		meshSpec      MeshSpec
+		fakeClientSet *fakeKubeClientSet
+		err           error
+	)
+	BeforeEach(func() {
+		meshSpec, fakeClientSet, err = bootstrapClient()
```

Comment:
guess there's no use of an `AfterEach` to clean up the client - we don't even have a cleanup

Reply:
I didn't think its necessary. The `BeforeEach` is to avoid test pollution by creating new clients which will result in new caches.